src/main/java/graphql/validation/rules/VariablesTypesMatcher.java

```diff
@@ -9,16 +9,38 @@
 import static graphql.schema.GraphQLNonNull.nonNull;
 import static graphql.schema.GraphQLTypeUtil.isList;
 import static graphql.schema.GraphQLTypeUtil.isNonNull;
+import static graphql.schema.GraphQLTypeUtil.unwrapNonNull;
 import static graphql.schema.GraphQLTypeUtil.unwrapOne;
 
 @Internal
 public class VariablesTypesMatcher {
 
-    public boolean doesVariableTypesMatch(GraphQLType variableType, Value variableDefaultValue, GraphQLType expectedType) {
-        return checkType(effectiveType(variableType, variableDefaultValue), expectedType);
+    /**
+     * This method and variable naming was inspired from the reference graphql-js implementation
+     *
+     * @param varType              the variable type
+     * @param varDefaultValue      the default value for the variable
+     * @param locationType         the location type where the variable was encountered
+     * @param locationDefaultValue the default value for that location
+     *
+     * @return true if the variable matches ok
+     */
+    public boolean doesVariableTypesMatch(GraphQLType varType, Value<?> varDefaultValue, GraphQLType locationType, Value<?> locationDefaultValue) {
+        if (isNonNull(locationType) && !isNonNull(varType)) {
+            boolean hasNonNullVariableDefaultValue =
+                    varDefaultValue != null && !(varDefaultValue instanceof NullValue);
+            boolean hasLocationDefaultValue = locationDefaultValue != null;
+            if (!hasNonNullVariableDefaultValue && !hasLocationDefaultValue) {
+                return false;
+            }
+            GraphQLType nullableLocationType = unwrapNonNull(locationType);
+            return checkType(varType, nullableLocationType);
+        }
+        return checkType(varType, locationType);
```

Comment:
if you looking in the graphql-js reference implementation - the logic is almost exactly the same